MDN React tutorials #1: Getting started (#30177)
mdn/content: files/en-us/learn/tools_and_testing/client-side_javascript_frameworks/react_getting_started/index.md

## Setting up your first React app

There are many ways to create a new React application, but we're going to use Vite, as mentioned earlier, to create a new application in our command line.

> It's possible to [add React to an existing project](https://react.dev/learn/add-react-to-an-existing-project) by copying some [`<script>`](/en-US/docs/Web/HTML/Element/script) elements into an HTML file, but using vite will allow tiy to spend more time building your app and less time fussing with setup.

Reviewer: @mxmason Ah, something just struck me as I was reviewing this — some bits have been written "you/your" and some bits have been written "we/our". It looks like you wrote it so that the former is generally used in general discussions, and the latter is generally used during tutorial instruction sections.

The usual style on MDN tends to be "you/your", so it might make sense for you to review it with this in mind, and change bits that make sense.

I don't think we should worry about changing article titles, as that would require us to modify the sidebar macro, which would be annoying ;-)

> It's possible to [add React to an existing project](https://react.dev/learn/add-react-to-an-existing-project) by copying some [`<script>`](/en-US/docs/Web/HTML/Element/script) elements into an HTML file, but using vite will allow tiy to spend more time building your app and less time fussing with setup.

Author: Good point, @chrisdavidmills. I think I wrote the entire tutorial in this way 😵. If you think it's worthwhile to spend time cleaning up the way we (;-D) address the audience, I can do that, but I think it would require more changes to more parts of the tutorial than we previously anticipated. I wasn't expecting to need to update _most_ of the tutorial, beyond the parts that reference code from `create-react-app`

> It's possible to [add React to an existing project](https://react.dev/learn/add-react-to-an-existing-project) by copying some [`<script>`](/en-US/docs/Web/HTML/Element/script) elements into an HTML file, but using vite will allow tiy to spend more time building your app and less time fussing with setup.

Reviewer: Yeah, true. In which case, just disregard this comment and carry on as you were. If we later decide that we want to reexplore this, we can always do it in a subsequent update without needing your attention on it.

### Import statements

The `import` statements at the top of the file allow `App.jsx` to use code that has been defined elsewhere. Let's look at these statements more closely.

```jsx
import { useState } from "react";
import reactLogo from "./assets/react.svg";
import viteLogo from "/vite.svg";
import "./App.css";
```

The first statement imports the `useState` hook from the `react` library. Hooks are a way of using React's features inside a component. We'll talk more about hooks later in this tutorial.

After that, we import `reactLogo` and `viteLogo`. Note that their import paths start with `./` and `/` respectively and that they end with the `.svg` extension at the end. This tells us that these imports are _local_, referencing our own files rather than npm packages.

The final statement imports the CSS related to our `App` component. Note that there is no variable name and no `from` directive. This is called a [_side-effect import_](/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only) — it doesn't import any value into the JavaScript file, but it tells Rollup, the bundler, to add the referenced CSS file to the final CSS bundle.

> ### The App component

Reviewer: ```suggestion
### The `App()` function
```

Reviewer: Also, you should do a search and replace — whenever you are talking specifically about the App function, you should write it with parens — `App()`

## Setting up your first React app

There are many ways to create a new React application. We're going to use Vite to create a new application via the command line.

It's possible to [add React to an existing project](https://react.dev/learn/add-react-to-an-existing-project) by copying some [`<script>`](/en-US/docs/Web/HTML/Element/script) elements into an HTML file, but using Vite will allow you to spend more time building your app and less time fussing with setup.

### Requirements

> In order to use Vite, you need to have [Node.js](https://nodejs.org/en/) installed. As of Vite 5.0, at least Node version 18 or later is required, and it's a good idea to run the latest long term support (LTS) version when you can. As of 24th October 2023, Node 20 is the latest LTS version. Node includes npm (the Node package manager).

Author: Not super thrilled with this section. I want to say more about Node, but it was also hard to identify the right level of detail before it gets way too out-of-scope for this tutorial.

> In order to use Vite, you need to have [Node.js](https://nodejs.org/en/) installed. As of Vite 5.0, at least Node version 18 or later is required, and it's a good idea to run the latest long term support (LTS) version when you can. As of 24th October 2023, Node 20 is the latest LTS version. Node includes npm (the Node package manager).

Author: Not super thrilled with this section on Node I want to say more, but it was also hard to identify the right level of detail before it gets way too out-of-scope for this tutorial.

Reviewer: I think this level of detail is fine.

If you're using Windows, you will need to install some software to give you parity with Unix/macOS terminal in order to use the terminal commands mentioned in this tutorial. **Gitbash** (which comes as part of the [git for Windows toolset](https://gitforwindows.org/)) or **[Windows Subsystem for Linux](https://docs.microsoft.com/windows/wsl/about)** (**WSL**) are both suitable. See [Command line crash course](/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line) for more information on these, and on terminal commands in general.

Also bear in mind that React and ReactDOM produce apps that only work on a fairly modern set of browsers — IE9+ by way of some polyfills. It is recommended that you use a modern browser like Firefox, Microsoft Edge, Safari, or Chrome when working through these tutorials.

See the following for more information:

- ["About npm" on the npm blog](https://docs.npmjs.com/about-npm/)
- ["Introducing npx" on the npm blog](https://blog.npmjs.org/post/162869356040/introducing-npx-an-npm-package-runner)
- [Vite's documentation](https://vitejs.dev/guide/)

### Initializing your app

> The npm package manager comes with a `create` command that allows you to create new projects from templates. We can use it to create a new app from Vite's standard React template. Make sure you `cd` to the place you'd like your app to live on your machine, then run the following in your terminal:

Author: I replaced your suggestion of "hard drive" with "machine" here. Technically, very few computing devices these days have hard drives ;)

Reviewer: Aye, makes sense ;-)

The `package.json` file is an npm feature that contains metadata about our project. This file is not unique to React applications; Vite merely populated it for us. You don't need to understand this file at all to complete this tutorial. However, if you'd like to learn more about it, you can read [package.json on the npm blog](https://docs.npmjs.com/cli/v9/configuring-npm/package-json/); we also talk about it in our [Package management basics](/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management) tutorial.

### Customizing our dev script

Before we move on, you might want to change your `package.json` file a little bit so that you don't have to pass the `--open` and `--port` flags every time you run `npm run dev`. Open `package.json` in your text editor and find the `scripts` object. Change the `"dev"` key so that it looks like this:

```diff
- "dev": "vite",
+ "dev": "vite --open --port 3000",
```

With this in place, your app will open in your browser at `http://localhost:3000` every time you run `npm run dev`.

> > **Note:** You _don't_ need the extra `--` here because we're passing arguments directly to `vite`, rather than to a pre-defined npm script.

Author: 🔍 How's this for the explanation about why `--` isn't necessary?

Reviewer: Yup, perfect!

Reviewer: Screenshot looks good too.

> ## Exploring our first React component — `<App />`

Author: 🔍  I decided to go with placing `<App />` in a code element for consistency. I don't _think_ I missed any other instances of things that belong in code.

Reviewer: I did spot a couple; commented.

├── package.json
├── public
│   └── vite.svg
├── src
│   ├── App.css
│   ├── App.jsx
│   ├── assets
│   │   └── react.svg
│   ├── index.css
│   └── main.jsx
└── vite.config.js
```

**`index.html`** is the most important top-level file. Vite injects your code into this file so that your browser can run it. You won't need to edit this file during our tutorial, but you should change the text inside the [`<title>`](/en-US/docs/Web/HTML/Element/title) element in this file to reflect the title of your application. Accurate page titles are important for accessibility.

The **`public`** directory contains static files that will be served directly to your browser without being processed by Vite's build tooling. Right now, it only contains a Vite logo.

The **`src`** directory is where we'll spend most of our time, as it's where the source code for our application lives. You'll notice that JavaScript files in this directory end in the extension `.jsx`. This extension is necessary for any file that contains JSX – it tells Vite to turn the JSX syntax into JavaScript that your browser can understand. The `src/assets` directory contains the React logo you saw in the browser.

> The `package.json` and `package-lock.json` files contain metadata about our project. These files are not unique to React applications: Vite populated `package.json` for us, and npm created `package-lock.json` for when we installed the app's dependencies. You don't need to understand these files at all to complete this tutorial. However, if you'd like to learn more about them, you can read about [`package.json`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json/) and [`package-lock.json`](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json) in the npm docs. We also talk about `package.json` in our [Package management basics](/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management) tutorial.

Author: These paragraphs are a little more complete. I mention the `assets` directory within `src` and acknowledge the `package.json` lockfile, with some links.

Reviewer: Looking good.

      </p>
    </>
  );
}

export default App;
```

The `App.jsx` file consists of three main parts: some [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statements at the top, the `App()` funtion in the middle, and an [`export`](/en-US/docs/Web/JavaScript/Reference/Statements/export) statement at the bottom. Most React components follow this pattern.

### Import statements

The `import` statements at the top of the file allow `App.jsx` to use code that has been defined elsewhere. Let's look at these statements more closely.

```jsx
import { useState } from "react";
import reactLogo from "./assets/react.svg";
import viteLogo from "/vite.svg";
import "./App.css";
```

The first statement imports the `useState` hook from the `react` library. Hooks are a way of using React's features inside a component. We'll talk more about hooks later in this tutorial.

After that, we import `reactLogo` and `viteLogo`. Note that their import paths start with `./` and `/` respectively and that they end with the `.svg` extension at the end. This tells us that these imports are _local_, referencing our own files rather than npm packages.

> The final statement imports the CSS related to our `<App />` component. Note that there is no variable name and no `from` directive. This is called a [_side-effect import_](/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only) — it doesn't import any value into the JavaScript file, but it tells Vite to add the referenced CSS file to the final code output, so that it can be used in the browser.

Author: Rather than explain yet more things (not just Rollup, but also what a bundler _is_), I've opted for this instead. I'm happy to mention Rollup by name and explain the concept of a bundler if you feel strongly about it.

Reviewer: Nope, this is fine. I think you've made a good choice here. It isn't necessary to go into bundlers at all.

```jsx
export default App;
```

This export statement makes our `App()` function available to other modules. We'll talk more about this later.

## Moving on to `main`

Let's open `src/main.jsx`, because that's where the `App()` function is being used. This file is the entry point for our app, and it initially looks like this:

```jsx
import React from "react";
import ReactDOM from "react-dom/client";
import App from "./App.jsx";
import "./index.css";

ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
);
```

As with `App.jsx`, the file starts by importing all the JS modules and other assets it needs to run.

The first two statements import the `React` and `ReactDOM` libraries because they are referenced later in the file. We don't write a path or extension when importing these libraries because they are not local files. In fact, they are listed as dependencies in our `package.json` file. Be careful of this distinction as you work through this lesson!

We then import our `App()` function and `index.css`, which holds global styles that are applied to our whole app.

> We then call the `ReactDOM.createRoot()` function, which defines the root node of our application. This takes as an argument the DOM element inside which we want our React app to be rendered. In this case, that's the DOM element with an ID of `root`. Finally, we chain the `render()` method onto the `createRoot()` call, passing it the JSX expression that we want to render inside our root. By writing `<App />` as this JSX expression, we're telling React to call the `App()` _function_ and render it as a _component_ inside the root node.

Author: I replaced bare mentions of `App` with `App()` where appropriate, but because there is a blend of styles in this file, I thought this line might help to disambiguate `App()` (a function) from`<App />` (a component). Might be worth an additional note about how functions _are_ components, but I'm not sure how exactly you'd like to tackle that. The conceptual difference is subtle.

Reviewer: I did make a few tweaks to help this a bit, but I think what we have here is good. You might want to check my "Small App() related tweaks" commit just to make sure you are happy.

      </p>
    </>
  );
}

export default App;
```

The `App.jsx` file consists of three main parts: some [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statements at the top, the `App` component in the middle, and an [`export`](/en-US/docs/Web/JavaScript/Reference/Statements/export) statement at the bottom. Most React components follow this pattern.

### Import statements

The `import` statements at the top of the file allow `App.jsx` to use code that has been defined elsewhere. Let's look at these statements more closely.

```jsx
import { useState } from "react";
import reactLogo from "./assets/react.svg";
import viteLogo from "/vite.svg";
import "./App.css";
```

The first statement imports the `useState` hook from the `react` library. Hooks are a way of using React's features inside a component. We'll talk more about hooks later in this tutorial.

After that, we import `reactLogo` and `viteLogo`. Note that their import paths start with `./` and `/` respectively and that they end with the `.svg` extension at the end. This tells us that these imports are _local_, referencing our own files rather than npm packages.

> The final statement imports the CSS related to our `App` component. Note that there is no variable name and no `from` directive. This is called a [_side-effect import_](/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only) — it doesn't import any value into the JavaScript file, but it tells Vite to add the referenced CSS file to the final code output, so that it can be used in the browser.

Author: @chrisdavidmills Rather than explain yet more things (not just Rollup, but also what a bundler _is_), I've opted for this instead. I'm happy to mention Rollup by name and explain the concept of a bundler if you feel strongly about it.

The `package.json` file contains information about our project that Node.js/npm uses to keep it organized. This file is not unique to React applications; Vite merely populated it for us. You don't need to understand this file at all to complete this tutorial, however, if you'd like to learn more about it, you can read [package.json on the npm blog](https://docs.npmjs.com/cli/v9/configuring-npm/package-json/); we also talk about it in our [Package management basics](/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management) tutorial.

### Customizing our dev script

Before we move on, you might want to change your `package.json` file a little bit so that you don't have to pass the `--open` and `--port` flags every time you run `npm run dev`. Open `package.json` in your text editor and find the `scripts` object. Change the `"dev"` key so that it looks like this:

```diff
- "dev": "vite",
+ "dev": "vite -- --open --port 3000",
```

> With this in place, your app will open in your browser at `http://localhost:3000` every time you run `npm run dev`.

Author: Ah, yes. Sorry about that! It didn't work for you because of a subtle syntax difference, which I've now noted in the article: the extra `--` to separate commands and args is only necessary when running _npm_ commands; it is not necessary if you're calling a program and passing an arg directly into it. Your "dev" command should be `"dev": "vite --open --port 3000"`

> ## Exploring our first React component — \<App/>

Reviewer: Stills needs actioning

      </p>
    </>
  );
}

export default App;
```

The `App.jsx` file consists of three main parts: some [`import`](/en-US/docs/Web/JavaScript/Reference/Statements/import) statements at the top, the `App` component in the middle, and an [`export`](/en-US/docs/Web/JavaScript/Reference/Statements/export) statement at the bottom. Most React components follow this pattern.

### Import statements

The `import` statements at the top of the file allow `App.jsx` to use code that has been defined elsewhere. Let's look at these statements more closely.

```jsx
import { useState } from "react";
import reactLogo from "./assets/react.svg";
import viteLogo from "/vite.svg";
import "./App.css";
```

The first statement imports the `useState` hook from the `react` library. Hooks are a way of using React's features inside a component. We'll talk more about hooks later in this tutorial.

After that, we import `reactLogo` and `viteLogo` Note that their import paths start with `./` and `/` respectively, and that they end with the `.svg` extension at the end – this tells us that these imports are _local_, referencing our own files rather than NPM packages.

> The final statement imports the CSS related to our App component. Note that there is no variable name and no `from` directive. This is called a [_side-effect import_](/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only) — it doesn't import any value into the JavaScript file, but it tells Rollup, the bundler, to add the referenced CSS file to the final CSS bundle.

Reviewer: Still needs actioning.

We won't be using the rest of the code! Replace the contents of `App.jsx` with the following:

```jsx
import viteLogo from "/vite.svg";
import "./App.css";

function App() {
  return (
    <>
      <header>
        <img src={viteLogo} className="logo" alt="Vite logo" />
        <h1>Hello, World!</h1>
      </header>
    </>
  );
}

export default App;
```

## Reactivity with variables and props

Next, we'll use our JavaScript skills to get a bit more comfortable editing components and working with data in React. We'll talk about how variables are used inside JSX, and introduce props, which are a way of passing data into a component (which can then be accessed using variables).

### Variables in JSX

Let's practice rendering some content from variables in React. As a reminder, our `App` component looks like this:

```jsx
import viteLogo from "/vite.svg";
import "./App.css";

function App() {
  return (
    <>
      <header>
        <img src={viteLogo} className="logo" alt="Vite logo" />
        <h1>Hello, World!</h1>
      </header>
    </>
  );
}

export default App;
```

> Declare a new variable, `subject`, before the return statement and set its value to the string `'React'`. Like this:

Reviewer: Also, for tutorial steps, you might want to consider numbered and indenting. I think this is the style we've used elsewhere in the tutorial set.

```jsx
const subject = "React";
function App() {
  return (
    <>
      <header>
        <img src={viteLogo} className="logo" alt="Vite logo" />
        <h1>Hello, World!</h1>
      </header>
    </>
  );
}
```

Then, replace the word "World" in the `<h1>` element with `{subject}`. Like this:

```jsx
// code omitted for brevity
<h1>Hello, {subject}!</h1>
```

Save your file and check your browser. You should see "Hello, React!".

### Component props

A **prop** is any data passed into a React component. React props are comparable to HTML attributes: where HTML elements have attributes, React components have props. Props are written inside component calls, and use the same syntax as HTML attributes — `prop="value"`. In React, the flow of data is unidirectional: props can only be passed from Parent components down to Child components; and props are read-only.

Let's open `index.jsx` and give our `<App />` component its first prop.

> Add a prop of `subject` to the `<App />` component call, with a value of `Clarice`. When you are done, your code should look something like this:

Reviewer: ```suggestion
Add a prop of `subject` to the `<App />` component call, with a value of `Clarice`. When you are done, it should look something like this:
```

```jsx
ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>

>     <App subject="Clarice" />

Reviewer: Just show this line; no need to show the whole file.

  </React.StrictMode>,
);
```

Back in `App.jsx`, let's revisit the App function itself, which reads like this (with the `return` statement shortened for brevity):

```jsx

> function App() {

Reviewer: I don't think you need to show what it looks like before the change. Just show the after view of it.

function App() {
  const [count, setCount] = useState(0);

  return (
    <>
      <div>
        <a href="https://vitejs.dev" target="_blank">
          <img src={viteLogo} className="logo" alt="Vite logo" />
        </a>
        <a href="https://react.dev" target="_blank">
          <img src={reactLogo} className="logo react" alt="React logo" />
        </a>
      </div>
      <h1>Vite + React</h1>
      <div className="card">
        <button onClick={() => setCount((count) => count + 1)}>
          count is {count}
        </button>
        <p>
          Edit <code>src/App.jsx</code> and save to test HMR
        </p>
      </div>
      <p className="read-the-docs">
        Click on the Vite and React logos to learn more
      </p>
    </>
  );
}
```

The `App` function returns a JSX expression. This expression defines what your browser ultimately renders to the DOM.

Just under the `return` keyword is a special bit of syntax: `<>`. This is a [fragment](https://react.dev/docs/fragments). React components have to return a single JSX element, and fragments allow us to do that without rendering arbitrary `<div>`s in the browser. You'll see fragments in many React applications.

> ### The export statement

Reviewer: ```suggestion
### The `export` statement
```

In React, the flow of data is unidirectional: props can only be passed from parent components down to child components.

Let's open `main.jsx` and give our `<App />` component its first prop.

Add a prop of `subject` to the `<App />` component call, with a value of `Clarice`. When you are done, it should look something like this:

```jsx
<App subject="Clarice" />
```

Back in `App.jsx`, let's revisit the `App()` function. Change the signature of `App()` so that it accepts `props` as a parameter and log `props` to the console so you can inspect it. Also delete the `subject` const; we don't need it anymore. Your `App.jsx` file should look like this:

```jsx
function App(props) {
  console.log(props);
  return (
    // code omitted for brevity
  );
}
```

Save your file and check your browser. You'll see a blank background with no content. This is because we're trying to read a `subject` variable that's no longer defined. Fix this by commenting out the `<h1>Hello {subject}!</h1>` line for a moment.

> > **Note:** If your text editor understands how to parse JSX (most modern editors do!), you can use its built-in commenting shortcut – `Ctrl + /` (on Windows) or `Cmd + /` (on macOS) to create comments more quickly.

Reviewer: ```suggestion
> **Note:** If your code editor understands how to parse JSX (most modern editors do!), you can use its built-in commenting shortcut — `Ctrl + /` (on Windows) or `Cmd + /` (on macOS) — to create comments more quickly.
```

Save the file with that line commented out. This time, you should see your
"Click me!" button rendered by itself. If you open your browser's developer console, you'll see a message that looks like this:

```plain
Object { subject: "Clarice" }
```

The object property `subject` corresponds to the `subject` prop we added to our `<App />` component call, and the string `Clarice` corresponds to its value. Component props in React are always collected into objects in this fashion.

> Let's use this `subject` prop to fix the error in our app. Un-coment the `Hello, {subject}!` line and change it to `Hello, {props.subject}!`, then delete the `console.log()` statement. Your code should look like this:

Reviewer: ```suggestion
Let's use this `subject` prop to fix the error in our app. Uncomment the `<h1>Hello, {subject}!</h1>` line and change it to `<h1>Hello, {props.subject}!</h1>`, then delete the `console.log()` statement. Your code should look like this:
```

```jsx
function App(props) {
  return (
    <>
      <header>

>         <h1>Hello, {subject}</h1>

Reviewer: ```suggestion
        <h1>Hello, {props.subject}!</h1>
```

> **Note:** `<App />` is rendered inside a special `<React.StrictMode>` component. This component helps developers catch potential problems in their code.

You can read up on these React APIs, if you'd like:

- [ReactDOM.createRoot()](https://react.dev/reference/react-dom/client/createRoot)
- [React.StrictMode](https://react.dev/reference/react/StrictMode)

## Starting fresh

Before we start building our app, we're going to delete some of the boilerplate code that Vite provided for us.

First, as an experiment, change the [`<h1>`](/en-US/docs/Web/HTML/Element/Heading_Elements) element in `App.jsx` so that it reads "Hello, World!", then save your file. You'll notice that this change is immediately rendered in the development server running at `http://localhost:3000` in your browser. Bear this in mind as you work on your app.

We won't be using the rest of the code! Replace the contents of `App.jsx` with the following:

```jsx
import "./App.css";

function App() {
  return (
    <>
      <header>
        <h1>Hello, World!</h1>
      </header>
    </>
  );
}

export default App;
```

> ## Practice with JSX

Author: This section is now much more robust than before, and I think it moves through concepts in a better order if I do say so myself.

Reviewer: Looks fabulous. I'll work through it again and see if I've got any comments on the new sections.

The curly braces around `subject` are another feature of JSX's syntax. The curly braces tell React that we want to read the value of the `subject` variable, rather than render the literal string `"subject"`. You can put any valid JavaScript expression inside curly braces in JSX; React will evaluate it and render the _result_ of the expression as the final content. Following is a series of examples, with comments above explaining what each expression will render:

```jsx-nolint
{/* Hello, React :)! */}
<h1>Hello, {subject + ' :)'}!</h1>
{/* Hello, REACT */}
<h1>Hello, {subject.toUpperCase()}</h1>
{/* Hello, 4 */}
<h1>Hello, {2 + 2}!</h1>
```

Even comments in JSX are written inside curly braces! This is because comments, too, are technically JavaScript expressions. The `/* block comment syntax */` is necessary for your program to know where the comment starts and ends.

> **Note:** If your text editor understands how to parse JSX (most modern editors do!), you can use its built-in commenting shortcut – `Ctrl + /` (on Windows) or `Cmd + /` (on macOS) to create comments more quickly.

### Component props

> **Props** are a means of passing data into a React component. Their syntax is identical to that of attrributes, in fact: `prop="value"`. The distinction between them is subtle: attributes are passed into plain elements; props are passed into React components.

Author: @chrisdavidmills I just spent time talking about attributes in the context of JSX. I don't think bringing "HTML" back here is helpful; nor is HTML the only relevant language with attributes

> **Props** are a means of passing data into a React component. Their syntax is identical to that of attrributes, in fact: `prop="value"`. The distinction between them is subtle: attributes are passed into plain elements; props are passed into React components.

Reviewer: Good call; I've updated my change suggestion to omit the instances of HTML

Let's open `index.jsx` and give our `<App />` component its first prop.

Add a prop of `subject` to the `<App />` component call, with a value of `Clarice`. When you are done, it should look something like this:

```jsx
<App subject="Clarice" />
```

Back in `App.jsx`, let's revisit the `App` function itself. Change the signature of `App` so that it accepts `props` as a parameter, and delete the `subject` const. Just like any other function parameter, you can put `props` in a `console.log()` to print it to your browser's console. That'll look like this"

```jsx
function App(props) {
  console.log(props);
  return (
    // code omitted for brevity
  );
}
```

With this change, `subject` becomes undefined, so comment out the line `Hello, {subject}!` for now. Save your file and check your browser's JavaScript console. You should see something like this logged:

```plain
Object { subject: "Clarice" }
```

The object property `subject` corresponds to the `subject` prop we added to our `<App />` component call, and the string `Clarice` corresponds to its value. Component props in React are always collected into objects in this fashion.

> Now that `subject` is one of our props, let's utilize it in `App.jsx`. Un-comment the `Hello, {subject}!` line and change it to be `Hello, {props.subject}!` instead, then delete the `console.log()` statement. Your code should look like this:

Reviewer: ```suggestion
Now that `subject` is one of our props, let's use it in `App.jsx`. Uncomment the `Hello, {subject}!` line and change it to `Hello, {props.subject}!`, then delete the `console.log()` statement. Your code should look like this:
```

```jsx
function App(props) {
  const subject = props.subject;
  return (

>     <>

Reviewer: This code snippet doesn't match the text instructions. Get rid of the `const` and put the `props.subject` inside the `<h1>` instead.

      <header>
        <h1>Hello, {subject}</h1>
        <button type="button" className="primary">
          Click me!
        </button>
      </header>
    </>
  );
}
```

> When you save, the app should now greet you with "Hello, Clarice!". If you return to `index.jsx`, edit the value of `subject`, and save, your text will change.

Reviewer: Oop, `index.jsx` should be `main.jsx`. There are a couple of instances to change.

> For additional practice, you might try adding a `greeting` prop to the `<App />` component and using it alongside the `subject` prop.

Reviewer: ```suggestion
For additional practice, you could try adding an additional `greeting` prop to the `<App />` component call inside `main.jsx` and using it alongside the `subject` prop inside `App.jsx`.
```

```jsx-nolint
{/* Hello, React :)! */}
<h1>Hello, {subject + ' :)'}!</h1>
{/* Hello, REACT */}
<h1>Hello, {subject.toUpperCase()}</h1>
{/* Hello, 4 */}
<h1>Hello, {2 + 2}!</h1>
```

Even comments in JSX are written inside curly braces! This is because comments, too, are technically JavaScript expressions. The `/* block comment syntax */` is necessary for your program to know where the comment starts and ends.

### Component props

**Props** are a means of passing data into a React component. Their syntax is identical to that of attributes, in fact: `prop="value"`. The difference is that whereas attributes are passed into plain elements, props are passed into React components.

In React, the flow of data is unidirectional: props can only be passed from parent components down to child components.

Let's open `main.jsx` and give our `<App />` component its first prop.

Add a prop of `subject` to the `<App />` component call, with a value of `Clarice`. When you are done, it should look something like this:

```jsx
<App subject="Clarice" />
```

> Back in `App.jsx`, let's revisit the `App()` function. Change the signature of `App()` so that it accepts `props` as a parameter and log `props` to the console so you can inspect it. Also delete the `subject` const; we don't need it anymore. Your `App.jsx` file should look like this:

Author: 🔍 Everything from here to "Let's use this `subject` prop..." has been slightly edited, hopefully to make the steps and results of those steps a little clearer

> **Note:** If your text editor understands how to parse JSX (most modern editors do!), you can use its built-in commenting shortcut – `Ctrl + /` (on Windows) or `Cmd + /` (on macOS) to create comments more quickly.

Save the file with that line commented out. This time, you should see your
"Click me!" button rendered by itself. If you open your browser's developer console, you'll see a message that looks like this:

```plain
Object { subject: "Clarice" }
```

The object property `subject` corresponds to the `subject` prop we added to our `<App />` component call, and the string `Clarice` corresponds to its value. Component props in React are always collected into objects in this fashion.

Let's use this `subject` prop to fix the error in our app. Un-coment the `Hello, {subject}!` line and change it to `Hello, {props.subject}!`, then delete the `console.log()` statement. Your code should look like this:

```jsx
function App(props) {
  return (
    <>
      <header>
        <h1>Hello, {subject}</h1>
        <button type="button" className="primary">
          Click me!

>         </button>

Author: 🔍 You mentioned possibly deleting this, but I think leaving it is fine. In the next lesson, we promptly ask the reader to delete and replace the contents of this file (and others)

Reviewer: Agreed, leaving it in is fine.

}
```

> With this change, `subject` becomes undefined, so comment out the line `Hello, {subject}!` for now. Save your file and check your browser's JavaScript console. You should see something like this logged:

Reviewer: In JSX, it is immediately not so obvious how to comment out a line of code. Maybe tell the reader to just delete that line for now, or show how to do it?

Reviewer: It looks like this is still an issue in the new draft.

```plain
Object { subject: "Clarice" }
```

The object property `subject` corresponds to the `subject` prop we added to our `<App />` component call, and the string `Clarice` corresponds to its value. Component props in React are always collected into objects in this fashion.

> Now that `subject` is one of our props, let's utilize it in `App.jsx`. Change the `subject` constant so that, instead of defining it as the string `React`, you are reading the value of `props.subject` and delete the `console.log()` statement. Your code should look like this:

Reviewer: ```suggestion
Now that `subject` is one of our props, let's utilize it in `App.jsx`. Update your code so that you are reading the `subject` value from `props.subject` rather than the string literal we defined previously. You can also delete the `console.log()` statement. Your code should look like this:
```
